services/src/main/java/org/keycloak/services/resources/SessionCodeChecks.java

```diff
@@ -181,35 +179,26 @@ public AuthenticationSessionModel initialVerifyAuthSession() {
 
         }
 
-        // See if we are already authenticated and userSession with same ID exists.
-        UserSessionModel userSession = authSessionManager.getUserSessionFromAuthCookie(realm);
-
-        boolean authenticating = !CookieHelper.getCookieValue(session, AUTH_SESSION_ID).isEmpty();
+        // Otherwise just try to restart from the cookie
+        RootAuthenticationSessionModel existingRootAuthSession = authSessionManager.getCurrentRootAuthenticationSession(realm);
+        response = restartAuthenticationSessionFromCookie(existingRootAuthSession);
 
-        if (authenticating) {
-            // if there is an auth session, make sure the user is not yet authenticated
+        // if restart from cookie was not found check if the user is already authenticated
+        if (response.getStatus() != Response.Status.FOUND.getStatusCode()) {
```

Comment:
It looks more reliable than just checking the presence of auth cookie. However, I'll wait for @mposolda feedback to see if there is some corner case that we might be missing.

It is not so nice to make the decision based on status code but I don't have a better idea other than refactoring more code and making the fix a bit more complex.

Reply:
I think it's OK now. if the response is not found it means there is no restart cookie, so there is no login (or re-login) in progress. Only in that case the check for identity cookie and already authenticated is performed.

With the previous order, the check for identity cookie was done if there was an AUTH_SESSION_ID cookie, and that created the problem for the impersonation. If you check the test, it's not even an impersonation, the AUTH_SESSION_ID cookies are removed, that triggered the problem at the second re-login because a new AUTH_SESSION_ID was generated but the auth session was immediately removed. In general the issue is triggered if the AUTH_SESSION_ID cookie is invalid (wrong, invalid, expired or whatever).

Reply:
+1 to add some "Impersonation integration" test instead of current `testReloginWithAndWithoutAuthSessionCookie` . The small issue with the current test is, that it uses lots of "implementation details" of Keycloak (for example manually opening auth pages, manually removing cookies etc), which typical user won't do. However at the same time, I understand that sometimes it is quite hard to test "real integration" .

My preference would be also to have "real impersonation" test, but if you confirm that it is impossible/unreliable to test, I am fine with current automated test as a fallback.

Reply:
I have done something, but it's not perfect. We cannot use selenium directly because we need the admin console to really use impersonation and that needs a real web driver (chrome, firefox,...). The default one is not enough. I have followed the spirit in the `ImpersonationTest` and done a impersonation using an HttpClient, getting all the cookies, set the cookies in the selenium driver and test the SAML app. I did it in the SAML tests because a SAML client is needed. Let's see the tests but it works OK to me.

Reply:
One more thing. I have maintained the other test (modified a bit) to test the general situation when the AUTH_SESSION_ID cookie is invalid.